/*
 * Copyright DataStax, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";
option java_package = "ai.langstream.agents.grpc";
option java_multiple_files = true;


service AgentService {
  rpc process(stream ProcessorRequest) returns (stream ProcessorResponse) {}
}

message Value {
  int32 schemaId = 1;
  oneof type_oneof {
    bytes bytesValue = 2;
    bool booleanValue = 3;
    string stringValue = 4;
    int32 byteValue = 5;
    int32 shortValue = 6;
    int32 intValue = 7;
    int64 longValue = 8;
    float floatValue = 9;
    double doubleValue = 10;
    string jsonValue = 11;
    bytes avroValue = 12;
  }
}

message Header {
  string name = 1;
  Value value = 2;
}

message Schema {
  int32 schemaId = 1;
  bytes value = 2;
}

message Record {
  int64 recordId = 1;
  optional Value key = 2;
  optional Value value = 3;
  repeated Header headers = 4;
  string origin = 5;
  optional int64 timestamp = 6;
}

message ProcessorRequest {
  Schema schema = 1;
  repeated Record records = 2;
}

message ProcessorResponse {
    Schema schema = 1;
    repeated ProcessorResult results = 2;
}

message ProcessorResult {
  int64 recordId = 1;
  optional string error = 2;
  repeated Record records = 3;
}